### PR TITLE
Develop v2.0.0

### PR DESCRIPTION
**Pull request by observeralone**
v2.0.0
plugin mechanism
move old tracing, metric, forwarded, redirect and serviceName to plugin

### CODE REVIEW

**Review by Oseenix — APPROVED** [review]
LGTM, Matcher DSL can added latter today

File: doc/matcher-DSL.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        .hasSuperClass("java.lang.String")
+        .hasInterface("java.lang.Comparable")
+        .notInterface()
+        .bilder()
+
+```
+
+- Logical OR  
+`ClassMatcherBuilder` provides `or()` method which will finish current builder and start a new `ClassMatcherBuilder`.
+The provious builder will generete a `Left ClassMatcher` and the new builder will generate a 'Right ClassMatcher'.
```
Comment (zhao-kun) [inline doc/matcher-DSL.md:69]:
```suggestion
The previous builder will generate a `Left ClassMatcher` and the new builder will generate a 'Right ClassMatcher'.
```

File: doc/matcher-DSL.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
+# Matcher DSL
+- [Overview](#Overview)
+- [ClassMatcher](#ClassMatcher)
+- [MethodMatcher](#MethodMatcher)
+
+## Overview
+EaseAgent provides a Matcher DSL (Domain Specific Language) to define a `Points` implementation which is used to defined classes and their methods to be enhanced. 
+
+The design of this DSL is borrowed from bytebuddy's class and method matching DSL design. Why not use bytebuddy's DSL interface directly?
+- Decopule from bytebuddy, which is an excellent bytecode editing framework, but we still want to decouple the API interface from it.
```
Comment (zhao-kun) [inline doc/matcher-DSL.md:10]:
```suggestion
- Decouple from the ByteBuddy, Although the byteBuddy is an excellent bytecode framework, we provide dedicated DSL to developers via abstracting it, can easily migrate to a more excellent framework in the future.
```

File: doc/matcher-DSL.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        return ClassMatcher.builder()
+            .hasInterface("javax.servlet.Filter")
+            .or()
+            .hasSuperClass("javax.servlet.http.HttpServlet")
+            .build();
+    }
+```
+
+- Logical NOT  
+The `negate()` method of `ClassMatcherBuilder` uses `NegateClassMatcher` to wrap the original matcher generated by builder when no such method is applied.
```
Comment (zhao-kun) [inline doc/matcher-DSL.md:84]:
```suggestion
The `negate()` method of `ClassMatcherBuilder` uses `NegateClassMatcher` to wrap the original matcher generated by the builder when no such method is applied.
```

File: doc/matcher-DSL.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+When a class matches the orignal matcher, it is unmatched by the wrap matcher; conversely, it is matched
```
Comment (zhao-kun) [inline doc/matcher-DSL.md:85]:
```suggestion
When a class matches the original matcher, it is unmatched by the wrap matcher; conversely, it is matched
```

File: doc/matcher-DSL.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
```
Comment (zhao-kun) [inline doc/matcher-DSL.md:86]: Give an example here to describe how to use `negate()`

File: doc/matcher-DSL.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+- [MethodMatcher](#MethodMatcher)
+
+## Overview
+EaseAgent provides a Matcher DSL (Domain Specific Language) to define a `Points` implementation which is used to defined classes and their methods to be enhanced. 
+
+The design of this DSL is borrowed from bytebuddy's class and method matching DSL design. Why not use bytebuddy's DSL interface directly?
+- Decopule from bytebuddy, which is an excellent bytecode editing framework, but we still want to decouple the API interface from it.
+- Provide a cleaner and easier to use API for Agent scenarios.
+
+
```
Comment (zhao-kun) [inline doc/matcher-DSL.md:13]: Need an overview to descript how many matchers we have and what's the purpose of matchers. eg:

We have defined two matchers, there are `ClassMatcher` and `MethodMatcher`.

- The `ClassMatcher` is dedicated to ....
- The `MethodMatcher` is dedicated to ...

File: doc/matcher-DSL.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+```java
+    MethodMatcher.builder()
+        .named("getConnection")
+        .returnType("java.sql.Connection")
+        .build())
+
+```
+
+In additon, `MethodMatcherBuilder` provides `and()` method which will finish current builder and start a new `MethodMatcherBuilder`.
+The provious builder will generete a `Left MethodMatcher` and the new builder will generate a 'Right MethodMatcher'.
```
Comment (zhao-kun) [inline doc/matcher-DSL.md:173]:
```suggestion
The previous builder will generate a `Left MethodMatcher` and the new builder will generate a 'Right MethodMatcher'.
```

File: doc/matcher-DSL.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+        .returnType("void")
+        .negate()
+        .build()
+```
+
+
+- Logical OR  
+`MethodMatcherBuilder` provides `or()` method which will finish current builder and start a new `MethodMatcherBuilder`.
+
+The provious builder will generete a `Left MethodMatcher` and the new builder will generate a 'Right MethodMatcher'.
```
Comment (zhao-kun) [inline doc/matcher-DSL.md:192]:
```suggestion
The previous builder will generate a `Left MethodMatcher` and the new builder will generate a 'Right MethodMatcher'.
```

**Review by akwei — APPROVED** [review]

**Comment by Oseenix** [comment]
v2.0.0 support plugin mechanism

**Pull request merged**